### DISCUSSION TITLE
Why identity map is weak and how to deal with it?

### DISCUSSION THREAD

**Discussion by 0xD09CD196D188D0B0** [sqlalchemy/sqlalchemy#7246] [Usage Questions]
I got an unexpected behavior in my code that uses asynchronous SQLAlchemy ORM with asyncpg: I'm fetching an instance from the database on earlier stages of request processing and then trying to use it again in depths of request processing using `AsyncSession.get()` method; Unexpectedly, it gets loaded from the database again, when I expect it being loaded from session's identity map.

During the debugging, I found out that the instance randomly just disappears from the identity map. Firstly, I was suspecting maybe some rollback to savepoint does the thing, but after several hours of looking for the cause, I still had no clue what is it.

And then I found out that the identity map's type is `WeakInstanceDict`. The word "weak" gave me a clue. And when I simply assigned the instance to some dummy global variable, everything started working. So it appears that GC collects the instance from memory at some point in time and that's because it disappears from the identity map.

Why identity map is implemented to be weak and how to prevent its contents from being garbage collected? I want to re-use the instance from the identity map always instead of doing excessive queries to the database. Passing it from the outer levels of code into the depths would be painful.

**Comment by zzzeek** [comment]
this is an issue of basic memory management.  if someone has a Session and they iterate through a million objects, making some changes on chunks of 1000 at a time and committing, then they lose all references to that chunk of objects, there's no reason at all for the Session to hold onto them, that would be (and certainly was, many years ago) reported as a memory leak.  

discussion on the weak referencing map is at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#is-the-session-a-cache .

we definitely had a recipe for a strong referencing identity map somewhere using the ``load`` event, there's https://github.com/sqlalchemy/sqlalchemy/wiki/SessionIndexing but that is too complicated for waht you need.   you can strong reference objects like this:

```py
from sqlalchemy.orm import Session
from sqlalchemy.orm import Mapper

def _index_object(session, instance):
    set_ = session.info.get("strong_set", None)
    if not set_:
        session.info["strong_set"] = set_ = set()

    set_.add(instance)

@event.listens_for(Mapper, "load")
def object_loaded(instance, ctx):
    _index_object(ctx.session, instance)

@event.listens_for(Session, "after_attach")
def index_object(session, instance):
    _index_object(session, instance)
```


for async_session, make sure you [follow the docs](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#using-events-with-the-asyncio-extension) for correctly associating events with async sessions.

full sync POC for the above:

```py
import gc

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Mapper
from sqlalchemy.orm import Session
Base = declarative_base()

def _index_object(session, instance):
    set_ = session.info.get("strong_set", None)
    if not set_:
        session.info["strong_set"] = set_ = set()

    set_.add(instance)

@event.listens_for(Mapper, "load")
def object_loaded(instance, ctx):
    _index_object(ctx.session, instance)

@event.listens_for(Session, "after_attach")
def index_object(session, instance):
    _index_object(session, instance)

class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add_all([A(data=f'a{i}') for i in range(20)])
s.commit()

s.execute(select(A)).all()
gc.collect()

# all objects strongly referenced
assert len(s.identity_map) == 20
```

**Reply by 0xD09CD196D188D0B0** [reply]
okay, it makes sense then. Thank you very much for your reply.
